Cpp_test/lib/ipc-shm.cpp

```diff
@@ -0,0 +1,113 @@
+#include "ipc-shm.h"
+
+IPCShm::~IPCShm()
+{
+
+	if (pthread_condattr_destroy(&(this->cond_attr)) != 0)
+	{
+		perror("Error at pthread_condattr_destroy()\n");
+	}
+	
+	if (pthread_mutexattr_destroy(&(this->mutex_attr)) != 0)
+	{
+		perror("Error at pthread_mutexattr_destroy()\n");
+	}
+	if (this->shm_ptr != nullptr)
+	{
+		munmap(this->shm_ptr, this->shm_size_in_bytes);
+	}
+}
+
+void IPCShm::init_cond()
+{
+
+	if (pthread_condattr_init(&(this->cond_attr)) != 0)
+	{
+		throw std::runtime_error("Error at pthread_condattr_init()");
+	}
+
+	if (pthread_condattr_setpshared(&(this->cond_attr), PTHREAD_PROCESS_SHARED) != 0)
+	{
+		throw std::runtime_error("Error at pthread_condattr_setpshared()");
+	}
+
+	/* Destory the cond attributes object */
+	if (pthread_condattr_destroy(&(this->cond_attr)) != 0)
+	{
+		perror("Error at pthread_condattr_destroy()\n");
+	}
+
+	if (pthread_cond_init(&(this->shm_ptr->cond), &(this->cond_attr)) != 0)
+	{
+		throw std::runtime_error(static_cast<std::string>("ERROR: pthread_cond_init() send: ") + strerror(errno));
+	}
+
+	if (pthread_cond_init(&(this->shm_ptr->cond_re), &(this->cond_attr)) != 0)
+	{
+		throw std::runtime_error(static_cast<std::string>("ERROR: pthread_cond_init() receive: ") + strerror(errno));
+	}
+}
+
+void IPCShm::init_mutex()
+{
+
+	if (pthread_mutexattr_init(&(this->mutex_attr)) != 0)
```

Comment:
The effect of initializing an already initialized mutex attributes object is undefined, you should try to destroy it before to be sure that your program doesn't have UB.

Reply:
I thought that I should destroy cond and mutex before exiting. With your idea, I should destroy them before initializing, right?

Reply:
I think you confuse two things: mutexattr and mutex.

According to the documentation, init an existing mutex_attr is undefined behavior. So I suggest to destroy the mutexattr before init it (just to be sure).

Destroying a mutex attribute does not destroy the mutex (only the attribute). So to clean up purpose, at the end of the program, you should destroy the mutexes created. Inside the program you can destroy the mutex attribute if you don't need it anymore, just to release the memory (but if you do it, you shall not use it anymore).